openedx/platform-plugin-aspects
# feat: send course/block tags with the CourseOverview and XBlock sinks

Tags are stored on the course_block_json/block_data_json, serialized as a list of "tag name=tag value" strings.

If the course/block has no tags, its json "tags" value in the json will be null.

Closes https://github.com/openedx/openedx-aspects/issues/217

**Testing instructions**

Setup:

1. Install this branch in your tutor dev/local environment, e.g
   ```
   tutor config save -a OPENEDX_EXTRA_PIP_REQUIREMENTS="git+https://github.com/openedx/platform-plugin-aspects.git@jill/sync-tags-on-course-sink#egg=platform_plugin_aspects"
   tutor dev launch -I
   ```
1. Enable tagging and use of the Course Authoring MFE by enabling these waffle flags for "everyone":

   ```
   contentstore.new_studio_mfe.use_tagging_taxonomy_list_page
   contentstore.new_studio_mfe.use_new_home_page
   contentstore.new_studio_mfe.use_new_course_outline_page
   contentstore.new_studio_mfe.use_new_unit_page
   ```
1. Set up some sample taxonomies, orgs, and courses.

   E.g. I have mounted edx-platform, and so can do this:

   ```
   cd edx-platform
   git clone https://github.com/open-craft/taxonomy-sample-data
   cd taxonomy-sample-data
   ```

   Update generate.py to provide a valid USER_EMAIL and TAXONOMY_SAMPLE_PATH, e.g
   ```diff
   --- a/generate.py
   +++ b/generate.py
   @@ -200,12 +200,12 @@ def import_tarfile_in_course(tarfile_path, course_key, user_id):
   
    User = get_user_model()

   -USER_EMAIL = "edx@example.com"
   +USER_EMAIL = "superuser@example.com"

    user = User.objects.get(email=USER_EMAIL)

    # Set to path where repo was cloned, eg: /edx/src/taxonomy-sample-data
   -TAXONOMY_SAMPLE_PATH = "/openedx/taxonomy-sample-data"
   +TAXONOMY_SAMPLE_PATH = "/openedx/edx-platform/taxonomy-sample-data"
   ```

   Run the script in the CMS shell:
   ```
   tutor dev run cms bash
   app@2e5e9f0f4f4e:~/edx-platform$ ./manage.py cms shell < taxonomy-sample-data/generate.py
   ```

Testing:

1. Navigate to the Course Authoring MFE and locate a tagged course, e.g.

   http://apps.local.edly.io:2001/course-authoring/course/course-v1:SampleTaxonomyOrg1+STC1+2023_1
1. Navigate to a unit in this course, and click Manage Tags in the right-hand sidebar to Edit Tags and add tags to one or more units.
1. Publish this course.
1. Visit Superset as a superuser.
1. Check that the course_overviews sink contains tag data in its `course_data_json`, e.g.

   ```
   {
      "advertised_start": null, 
      "announcement": null,
      "lowest_passing_grade": 0.5,
      "invitation_only": false,
      "max_student_enrollments_allowed": null,
      "effort": null,
      "enable_proctored_exams": false,
      "entrance_exam_enabled": false,
      "external_id": null,
      "language": "en",
      "tags": "['HierarchicalTaxonomy=hierarchical taxonomy tag 2.5.14', 'HierarchicalTaxonomy=hierarchical taxonomy tag 3.13.28', 'HierarchicalTaxonomy=hierarchical taxonomy tag 3.6.45', 'ESDC Skills and Competencies=Building and Construction', 'Lightcast Open Skills Taxonomy=Stream Processing', 'FlatTaxonomy=flat taxonomy tag 2440']",
   }
   ```
1. Check that the course_blocks sink contains tag data in its `block_data_json`, e.g.

   ```
   {
      "course": "STC1",
      "run": "2023_1",
      "block_type": "about",
      "detached": 1,
      "graded": 0,
      "completion_mode": "",
      "section": 2,
      "subsection": 1,
      "unit": 5,
      "tags": "['ESDC Skills and Competencies=Endurance', 'ESDC Skills and Competencies=Physical Strength Abilities']",
   }
   ```

**Author Notes & Concerns**

1. <strike>There's a bug in the Course Authoring MFE that prevents the Manage Tags UI from saving tags for units (step 2 under Testing). Branch that fixes this: TBD.</strike> fixed by https://github.com/openedx/edx-platform/pull/34490.
2.  <strike>The Manage Tags UI shows a tag count that includes explicit + implicit tags. When tags from a hierarchical taxonomy are serialized for the course/block, only the explicit "leaf" tags are stored, so the counts will differ.</strike> [6055922](https://github.com/openedx/platform-plugin-aspects/pull/41/commits/6055922e75b5582f83f63adaf8fd244338503b92) ensures both implicit and explicit tags are serialized to ClickHouse.
3. Do we need to update the aspects migrations to add a `tags` field to the materialized views [like we do for other block_data_json fields](https://github.com/openedx/tutor-contrib-aspects/blob/main/tutoraspects/templates/aspects/apps/aspects/migrations/alembic/versions/0035_migrate_dictionaries.py#L178)? 


**Merge checklist:**
Check off if complete *or* not applicable:
